### DISCUSSION TITLE
Close dropdown menu on click outside of element

### DISCUSSION THREAD

**Discussion by bk973** [vercel/next.js#44515] [Help] [answered]
I am using a custom hook to track click event outside a specified element reference(ref)

`
import { useEffect, useRef } from "react";
import { toUSVString } from "util";

export function useClickOutside(callback: () => void) {
  const ref = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    function handleClick(event: MouseEvent) {
      const target = event.target as HTMLDivElement | HTMLButtonElement;
      if (event.target === ref.current) {
        return;
      }

      if (target?.contains(ref.current) && target !== ref.current) {
        console.log("clicked outside");
        callback();
      }
    }

    window && window.addEventListener("click", handleClick, { capture: true });
    return () => {
      window &&
        window.removeEventListener("click", handleClick, { capture: true });
    };
  }, [callback, ref.current]);

  return ref;
}
`
I am using a reusable component because I want to create multiple dropdowns through the entire site. I am also using context API to keep track of open dropdowns using unique id. This helps to prevent two dropdowns being opened at the same time. Here's the code for DropDown component.

`
import { ReactNode } from "react";
import { useClickOutside } from "../../util/hooks/useClickOutside";

export default function DropDown({
  id,
  onOpen,
  onClose,
  isOpen,
  content,
  buttonContent,
}: {
  id: string;
  onOpen: (id: string) => void | undefined;
  onClose: () => void | undefined;
  isOpen: boolean;
  content?: ReactNode;
  buttonContent?: any;
}) {

    let elementRef = useClickOutside(() => {
    if (isOpen) {
      onClose();
    }
  });
  

  return (
    <div className="relative z-30">
      <button
        className="rounded-full p-0 border-2 border-slate-300"
        onClick={() => (isOpen ? onClose() : onOpen(id))}
      >
        {buttonContent}
      </button>
      {isOpen && (
        <div
          ref={elementRef}
          className="absolute space-y-2 right-0 w-56 mt-2 origin-top-right bg-white rounded-lg shadow-xl ring-1 ring-black ring-opacity-5 focus:outline-none z-30"
        >
          {content}
        </div>
      )}
    </div>
  );
}
`

**Behavior**
DropDown closes when I click outside of the element, but within the component that wraps the dropdown. I am using a complex Layout consisting of side menu, sections and app bar. When I click on any other component where the dropdown isn't wrapped, the event is not fired.

**Expected Behavior**
I want to be able to close the dropdown when I click anywhere outside the element but within the entire page.

I have recorded a loom video with more explanations
[Loom Video](https://www.loom.com/share/e6de04b066c442899f588c9db076f161)

**Comment by icyJoseph** [comment, marked as answer]
Try changing your handler to:

```tsx
  function handleClick(event: MouseEvent) {
      const target = event.target as HTMLDivElement | HTMLButtonElement; // maybe cast to as Node or HTMLElement

      const el = ref.current;

      if (!el) return;

      if (event.target === ref.current) {
        return;
      }

      if (el.contains(event.target as Node)) {
        return;
      }

      console.log('clicked outside');
      callback();
    }
```

I do think however, that maybe you want something more like:

```tsx
import { RefObject, useEffect } from 'react';

type Event = MouseEvent | TouchEvent | KeyboardEvent;

const isKeyboardEvent = (event: Event): event is KeyboardEvent => {
  return 'key' in event;
};

export const useOnClickOutside = <T extends HTMLElement = HTMLElement>(
  ref: RefObject<T>,
  handler: (event: Event) => void
) => {
  useEffect(() => {
    const listener = (event: Event) => {
      const elem = ref.current;

      if (!elem) return;

      if (isKeyboardEvent(event)) {
        if (event.key === 'Escape') {
          handler(event);
        }
      }

      if (elem.contains(event.target as HTMLElement)) {
        return;
      }

      handler(event);
    };

    document.addEventListener('mousedown', listener);
    document.addEventListener('touchstart', listener);
    document.addEventListener('keydown', listener);

    return () => {
      document.removeEventListener('mousedown', listener);
      document.removeEventListener('touchstart', listener);
      document.removeEventListener('keydown', listener);
    };
  }, [ref, handler]);
};
```

And then you place the responsibility of the ref creation on the caller:

```tsx
import { useState, useRef } from 'react';
import { useOnClickOutside } from '../hooks/useOnClickOutside';

const Home = () => {
  const [open, setOpen] = useState(false);
  const ref = useRef<HTMLDivElement>(null);
  useOnClickOutside(ref, () => setOpen(false));

  return (
    <>
      <header>I am the header!</header>

      <button onClick={() => setOpen(true)}>Click me</button>
      <div style={{ height: '90vh', background: 'papayawhip' }}>
        <div
          ref={ref}
          style={{
            background: 'palevioletred',
            height: open ? '10vh' : 'auto',
          }}
        >
          {open ? 'open' : 'close'}
        </div>
      </div>

      <footer>I am the footer!</footer>
    </>
  );
};

export default Home;
```

**Reply by bk973** [reply]
Thank you so much @icyJoseph Your answer has been helpful. You have saved my day. Taking the responsibility of ref creation to the caller made the magic. Thanks again and again...
